### DISCUSSION TITLE
Constraints on associated links

### DISCUSSION THREAD

**Discussion by forsakendaemon** [geldata/gel#3301] [Q&A] [answered]
So I'm trying to model something like a multiple choice question and answer system. My first attempt looked something like this:

```
type Question {
    required property text -> str;
    property description -> str;
    required multi link choices -> Choice;
}

type Choice {
    required property text -> str;
    link questions := (.<choices[IS Question])
}

type Response {
    required link choice -> Choice;
    required link question -> Question;
}
```

My issue comes when trying to add a constraint to `Response` such that `.choice` is restricted to the values of `.question.choices`, as constraints are restricted to paths with only one hop. The `Choice`s themselves have meaning independent of the `Question`s, which is why I'm modelling them separately (rather than as an array within `Question` and a pointer into that array or something like that).

I'm sure I'm missing a really obvious way to refactor this to make more sense, but I've been staring at it for two days and just can't see it.

Thanks so much!

**Comment by vpetrovykh** [comment, marked as answer]
I'd like to clarify something about this question. There are 2 statements here that seem to be at odds:
1) "add a constraint to `Response` such that `.choice` is restricted to the values of `.question.choices`"
2) "The `Choice`s themselves have meaning independent of the `Question`s"

If the choice is meaningful, self-contained and independent of the question, then why does it need to be restricted _based_ on the question? If it needs to be restricted based on the question, then clearly it's not independent... 

On the practical side of things, in my experience the multiple choice questionnaires that satisfied the condition of choices being independent of the questions were the kind that had special machine-readable cards to go with them. The choices were fixed (something like "A", "B", "C", "D") and same for all questions. To model this, you might want to use an `enum`, something like this:

```
type Question {
    required property text -> str;
    property description -> str;
}

scalar type Choice extending enum<A, B, C, D>;

type Response {
    required property choice -> Choice;
    required link question -> Question;
}
```

If the point is that you want to make `Choice` configurable somehow, in terms of how many options there are, then you might do something like this instead:
```
type Question {
    required property text -> str;
    property description -> str;
    required property options -> array<str>;
}

type Choice {
    # this will be populated based on the question's "num_choices"
    required property num_choices -> int64;  
    property value -> int64;
    # The ?? 0 is to simply accept an empty value as valid
    constraint expression on (
        with v := .value ?? 0 
        select 0 <= v and v < .num_choices
    );
};

type Response {
    required link choice -> Choice;
    required link question -> Question;
}
```

The way to use it would be something like this:
```
db> insert Question{text := 'What is 2 + 2?', options := ['4', '3', '0']};
{default::Question {id: a6f32228-62ac-11ec-902f-d3010e1e4570}}
db> with Q := (select Question limit 1)
... insert Response {
...   question := Q,
...   choice := (insert Choice {num_choices := len(Q.options), value := 1})
... };
{default::Response {id: bbc2bae2-62ac-11ec-b360-67facd9d82c3}}
db> select Response {
...   question: {text},
...   choice: {value},
...   answer := .question.options[.choice.value]
... };
{
  default::Response {
    question: default::Question {text: 'What is 2 + 2?'},
    choice: default::Choice {value: 1},
    answer: '3',
  },
}
```
The connection between the `Choice` and `Question` is only made when creating the `Response`. There's nothing that would allow to extend the constraint across multiple hops in a path (so can't be spread across multiple objects). Typically you control the point where the data is written and are able to enforce `num_choices := len(Q.options)` while taking in the user's input for `value`.

**Reply by forsakendaemon** [reply]
Mmmmm, this is a way I looked at trying to make this work originally.

So essentially while the choices have meaning independent of the question, they don't have meaning in every question. For example, if one question is "If given $5, which of these would you buy? (Lottery ticket / loaf of bread / chocolate bar)" then some of those choices can transfer over to "Which would you most like to have for dinner?", but the lottery ticket probably wouldn't work. Sharing choices between them gives information about preferences between loaves of bread and chocolate bars, though, which is what I'm eventually looking into, along with the structure of the bipartite graph of questions and choices (as I'm trying to calculate informational density and things like that).

I come from a social science research background, so most of my experience  with questions like this also involved fixed response structures, if not fixed response options. With so much data collection being done online now, though, I'm getting some much more weird data that I'm now trying to unify into a single queriable whole! I would not have designed it like this, believe me.

I think that maybe making the data model more flexible (i.e. allowing any choice to go with any question) and then enforcing any constraints exclusively on the client side might be the best way forward here. Which is frustrating, but it will work!

Thanks so much for taking the time to respond - your answer was clear and helped this make a whole lot more sense to me.

**Reply by vpetrovykh** [reply]
Interesting. So if the goal is to allow relative freedom in defining choice options and then actually sharing some of these objects across different questions (and responses), then perhaps this could be a useful querying trick (using your original schema):
```
select Response {
  question: {text},
  choice := (select C := .choice filter C in Response.question.choices) {text}
}
```
What this query does is filtering out any `choice` if it wasn't among the available `choices` for the specific question. It doesn't stop you from _writing_ an odd choice into a `Response`, but it can tell you if you did so. If you want to prevent this from being written in the first place, I'd recommend this:
```
with 
  Q := (select Question filter .id = <uuid>$qid)
insert Response {
  question := Q,
  choice := (select Q.choices filter .id = <uuid>$cid),
}
```
Where `$qid` and `$cid` are the question and choice IDs passed in as parameters, respectively. The above query will fail if you pass a `Choice` ID that's not part of the available choices for the question, so it should do the enforcement you want.

When `Choice` objects are actually shared among questions, not merely when several `Choice` objects have identical text, this tactic will allow then to find _responses_ that share a given `Choice`. You might want to add `constraint exclusive` to the `Choice` `text` property if every unique option text represents the same choice conceptually.
